src/mono/wasm/Wasm.Build.Tests/IcuShardingTests.cs

```diff
@@ -0,0 +1,87 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.IO;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+using System.Collections.Generic;
+
+#nullable enable
+
+namespace Wasm.Build.Tests;
+
+public class IcuShardingTests : BuildTestBase
+{
+    public IcuShardingTests(ITestOutputHelper output, SharedBuildPerTestClassFixture buildContext)
+        : base(output, buildContext)
+    {
+    }
+
+    public static IEnumerable<object?[]> IcuExpectedAndMissingShardTestData(bool aot, RunHost host)
+        => ConfigWithAOTData(aot)
+            .Multiply(
+                new object[] { "icudt.dat", new string[] { "en-GB", "zh-CN", "hr-HR" }, new string[] { "xx-yy" } },
+                new object[] { "icudt_EFIGS.dat", new string[] { "en-US", "fr-FR", "es-ES" }, new string[] { "pl-PL", "ko-KR", "cs-CZ" } },
+                new object[] { "icudt_CJK.dat", new string[] { "en-GB", "zh-CN", "ja-JP" }, new string[] { "fr-FR", "hr-HR", "it-IT" } },
+                new object[] { "icudt_no_CJK.dat", new string[] { "en-AU", "fr-FR", "sk-SK" }, new string[] { "ja-JP", "ko-KR", "zh-CN"} })
+            .WithRunHosts(host)
+            .UnwrapItemsAsArrays();
+
+    private static string GetProgramText(string[] expectedLocales, string[] missingLocales) => $@"
+        using System;
+        using System.Globalization;
+
+        string[] expectedLocales = new string[] {{ "" {string.Join("\", \"", expectedLocales)} "" }};
+        string[] missingLocales =  new string[] {{ "" {string.Join("\", \"", missingLocales)} "" }};
+        foreach (var loc in expectedLocales)
+        {{
+            var culture = new CultureInfo(loc);
+            Console.WriteLine($""Found expected locale: {{loc}} - {{culture.Name}}"");
+        }}
+        foreach (var loc in missingLocales)
+        {{
+            try
+            {{
+                var culture = new CultureInfo(loc);
+            }}
+            catch(Exception)
+            {{
+            Console.WriteLine($""Missing locale as planned: {{loc}}"");
+            }}
+        }}
+        return 42;
+        ";
+
+    // on Chrome: when loading only EFIGS or only CJK, CoreLib's failure on culture not found cannot be easily caught:
+    // Encountered infinite recursion while looking up resource 'Argument_CultureNotSupported' in System.Private.CoreLib.
+    [Theory]
```

Comment:
Other tests to add:

- test with the `WasmBundleFullIcuData=true` combined with, and without InvariantGlobalization
- test with a custom icudt file, and loaded via the WasmIcuDataFileName property
- test with a custom icudt file, and loaded via the WasmIcuDataFileName property + WasmBundleFullIcuData property set

Reply:
ad. dot 2) Could you help me figure out how to test a custom icudt file that we are not shipping? Once I push something like that, it will always fail because that file won't exist. Even if I create a dummy .dat file in the test (I mean I would start the test, open a file stream, save a new file somewhere on the machine, then build and run the project) then it won't have an ICU datafile structure, so I won't really test if loading data works. Only that we loaded it. Do you have a better idea than this one?

Reply:
You can generate a valid one and add in `src/mono/wasm/testassets`.

Reply:
.. then always load that for the test.